### PR TITLE
Refactor logic to allow for UI-dependent behaviour

### LINKED ISSUE #81: Refactor logic to allow for UI-dependent behaviour
Currently, the UI merely passes the user input to the logic component for the input to be parsed and executed. However, this does not allow behaviour to be dependent on the current state of the UI. For example, we want certain commands to behave differently depending on whether we are on the expense list or the income list.

To add a data class `UiState` that captures all the important information of the current state of the UI that is passed to the logic component.

### PR DESCRIPTION
**Pull request by ianyong**
Resolves #81.

### CODE REVIEW

File: src/main/java/ay2021s1_cs2103_w16_3/finesse/logic/Logic.java [hunk 1]
```diff
@@ -18,12 +19,14 @@
 public interface Logic {
     /**
      * Executes the command and returns the result.
+     *
      * @param commandText The command as entered by the user.
+     * @param uiState The current state of the UI.
      * @return the result of the command execution.
```
Comment (zhaojj2209) [inline src/main/java/ay2021s1_cs2103_w16_3/finesse/logic/Logic.java:25]:
```suggestion
     * @return The result of the command execution.
```

File: src/main/java/ay2021s1_cs2103_w16_3/finesse/ui/UiState.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+     */
+    public UiState(Tab currentTab) {
+        requireNonNull(currentTab);
+        this.currentTab = currentTab;
+    }
+
+    /**
+     * Returns the currently selected tab.
+     *
+     * @return the currently selected tab.
```
Comment (zhaojj2209) [inline src/main/java/ay2021s1_cs2103_w16_3/finesse/ui/UiState.java:44]:
```suggestion
     * @return The currently selected tab.
```

File: src/main/java/ay2021s1_cs2103_w16_3/finesse/ui/UiState.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+     */
+    public Tab getCurrentTab() {
+        return currentTab;
+    }
+
+    /**
+     * Updates the currently selected tab.
+     *
+     * @param currentTab the currently selected tab.
```
Comment (zhaojj2209) [inline src/main/java/ay2021s1_cs2103_w16_3/finesse/ui/UiState.java:53]:
```suggestion
     * @param currentTab The currently selected tab.
```

**Review by zhaojj2209 — CHANGES_REQUESTED** [review]
Just some changes to capitalisation for consistency.

**Review by zhaojj2209 — APPROVED** [review]
LGTM!

File: src/main/java/ay2021s1_cs2103_w16_3/finesse/logic/parser/FinanceTrackerParser.java [hunk 4]
```diff
@@ -34,11 +35,12 @@
     /**
      * Parses user input into command for execution.
      *
-     * @param userInput full user input string
-     * @return the command based on the user input
-     * @throws ParseException if the user input does not conform the expected format
+     * @param userInput Full user input string.
+     * @param uiState Current state of the UI.
+     * @return The command based on the user input.
+     * @throws ParseException If the user input does not conform the expected format.
```
Comment (yongping827) [inline src/main/java/ay2021s1_cs2103_w16_3/finesse/logic/parser/FinanceTrackerParser.java:41]:
```suggestion
     * @throws ParseException If the user input does not conform to the expected format.
```

File: src/main/java/ay2021s1_cs2103_w16_3/finesse/ui/UiState.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+ * Represents the state of the current UI.
+ */
+public class UiState {
+    /**
+     * Tabs present in the UI.
+     */
+    public enum Tab {
+        OVERVIEW,
+        INCOME,
+        EXPENSES,
```
Comment (yongping827) [inline src/main/java/ay2021s1_cs2103_w16_3/finesse/ui/UiState.java:15]: Either
```suggestion
        INCOMES,
        EXPENSES,
```
or
```suggestion
        INCOME,
        EXPENSE,
```

File: src/test/java/ay2021s1_cs2103_w16_3/finesse/logic/parser/FinanceTrackerParserTest.java [hunk 6]
```diff
@@ -33,102 +33,501 @@
...
+    @Test
+    public void parseCommand_listIncomeWhenAnalyticsTab() throws Exception {
+        assertTrue(parser.parseCommand(ListIncomeCommand.COMMAND_WORD, overviewUiStateStub)
+                instanceof ListIncomeCommand);
+        assertTrue(parser.parseCommand(ListIncomeCommand.COMMAND_WORD + " 3",
+                analyticsUiStateStub) instanceof ListIncomeCommand);
+    }
+
+    @Test
+    public void parseCommand_unrecognisedInputWhenOverviewTab_throwsParseException() {
```
Comment (yongping827) [inline src/test/java/ay2021s1_cs2103_w16_3/finesse/logic/parser/FinanceTrackerParserTest.java:432]:
```suggestion
    public void parseCommand_unrecognizedInputWhenOverviewTab_throwsParseException() {
```
Also in other similar places.

**Review by yongping827 — CHANGES_REQUESTED** [review]

**Review by yongping827 — APPROVED** [review]
LGTM

**Pull request merged**